lib/Runtime/Base/ThreadContext.h

```diff
@@ -99,24 +99,28 @@ class InterruptPoller
             void * __frameAddr = nullptr; \
             GET_CURRENT_FRAME_ID(__frameAddr); \
             Js::LeaveScriptObject<stackProbe, leaveForHost, isFPUControlRestoreNeeded> __leaveScriptObject(scriptContext, __frameAddr); \
+            BEGIN_NO_EXCEPTION; \
             AutoReentrancyHandler autoReentrancyHandler(scriptContext->GetThreadContext());
 #define LEAVE_SCRIPT_END_EX(scriptContext) \
             if (scriptContext != nullptr) \
                 {   \
                     scriptContext->GetThreadContext()->DisposeOnLeaveScript(); \
                 }\
+            END_NO_EXCEPTION; \
         }
 
-#define LEAVE_SCRIPT_IF(scriptContext, condition, block) \
-        if (condition) \
+#define LEAVE_SCRIPT_IF_ACTIVE(scriptContext, active, externalCall) \
+        if (active) \
         { \
             BEGIN_LEAVE_SCRIPT(scriptContext); \
-            block \
+            externalCall \
             END_LEAVE_SCRIPT(scriptContext); \
         } \
         else \
         { \
-            block \
+            BEGIN_NO_EXCEPTION; \
```

Comment:
@curtisman what are the DOM callout points where we can have exceptions even if we've left script?

Reply:
Is `LEAVE_SCRIPT_IF`/`LEAVE_SCRIPT_IF_ACTIVE` only used in the SourceTextModuleRecord file, for notifying the host that we have parsed a JS module? That seems like a much more constrained scenario than calling out to the DOM in general.

Reply:
The goal is to start using this macro in DynamicSourceHolder. Moving LEAVE_SCRIPT_START_EX to BEGIN_NO_EXCEPTION was done on the assumption that there cannot be any exceptions across the external call boundary (which, according to Curtis, isn't always true, so I'm synchronizing LEAVE_SCRIPT_START_EX and LEAVE_SCRIPT_IF_ACTIVE to use SAVE_EXCEPTION_CHECK).

---
In reply to: [196641408](https://github.com/Microsoft/ChakraCore/pull/5335#discussion_r196641408) [](ancestors = 196641408)